cluster-autoscaler/processors/provreq/provisioning_request_injector.go

```diff
@@ -0,0 +1,91 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provreq
+
+import (
+	"time"
+
+	apiv1 "k8s.io/api/core/v1"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/autoscaler/cluster-autoscaler/context"
+	"k8s.io/autoscaler/cluster-autoscaler/processors/pods"
+	"k8s.io/autoscaler/cluster-autoscaler/provisioningrequest/apis/autoscaling.x-k8s.io/v1beta1"
+	provreqconditions "k8s.io/autoscaler/cluster-autoscaler/provisioningrequest/conditions"
+	provreqpods "k8s.io/autoscaler/cluster-autoscaler/provisioningrequest/pods"
+	"k8s.io/autoscaler/cluster-autoscaler/provisioningrequest/provreqclient"
+	"k8s.io/client-go/rest"
+	"k8s.io/klog/v2"
+)
+
+const (
+	defaultRetryTime = 10 * time.Minute
+)
+
+// ProvisioningRequestPodsInjector creates in-memory pods from ProvisioningRequest and inject them to unscheduled pods list.
+type ProvisioningRequestPodsInjector struct {
+	client provisioningRequestClient
+	now    func() time.Time
+}
+
+// Process pick one ProvisioningRequest, update Accepted condition and inject pods to unscheduled pods list.
+func (p *ProvisioningRequestPodsInjector) Process(
+	_ *context.AutoscalingContext,
+	unschedulablePods []*apiv1.Pod,
+) ([]*apiv1.Pod, error) {
+	provReqs, err := p.client.ProvisioningRequests()
+	if err != nil {
+		return nil, err
+	}
+	for _, pr := range provReqs {
+		conditions := pr.Conditions()
+		if apimeta.IsStatusConditionTrue(conditions, v1beta1.Failed) || apimeta.IsStatusConditionTrue(conditions, v1beta1.Provisioned) {
+			continue
+		}
+
+		provisioned := apimeta.FindStatusCondition(conditions, v1beta1.Provisioned)
+		accepted := apimeta.FindStatusCondition(conditions, v1beta1.Provisioned)
+
+		//TODO(yaroslava): support exponential backoff
+		// Inject pods if ProvReq is new or it has Provisioned == False condition more than defaultRetryTime
+		if accepted == nil || provisioned != nil && provisioned.Status == metav1.ConditionFalse &&
+			provisioned.LastTransitionTime.Add(defaultRetryTime).Before(p.now()) {
+			provreqpods, err := provreqpods.PodsForProvisioningRequest(pr)
+			if err != nil {
+				klog.Errorf("Failed to get pods for ProvisioningRequest %v", pr.Name())
+				provreqconditions.AddOrUpdateCondition(pr, v1beta1.Failed, metav1.ConditionTrue, provreqconditions.FailedToCreatePodsReason, err.Error(), metav1.NewTime(p.now()))
+				continue
+			}
+			unschedulablePods := append(unschedulablePods, provreqpods...)
+			provreqconditions.AddOrUpdateCondition(pr, v1beta1.Accepted, metav1.ConditionTrue, provreqconditions.AcceptedReason, provreqconditions.AcceptedMsg, metav1.NewTime(p.now()))
```

Comment:
Maybe I'm missing some context, but why would we have Status be updated to "Accepted" when the check is done as `accepted == nil`.
Is it a transitive which you're setting here, that will change the flag to `accepted != nil`

Reply:
`Accepted indicates that the ProvisioningRequest was accepted by ClusterAutoscaler, so ClusterAutoscaler will attempt to provision the nodes for it.`
If Accepted == nil, means CA see ProvReq first time and should set this condition, since the pods from ProvReq are injected.